### DISCUSSION TITLE
use cgltf or tinygltf as a no-assimp compile mode for meshIO?

### DISCUSSION THREAD

**Discussion by starseeker** [elalish/manifold#1010] [General]
Was thinking a little about meshIO - it's very handy to have when one needs to write out example "bad" inputs for debugging, but needing assimp complicates life a bit.  On the other hand, using it does provide support for a lot of formats.

What would the devs say to using something really small and embed-able like cgltf or tinygltf to support building meshIO when assimp isn't available, and using assimp instead when it is?  Essentially that would allow for an "always on" aspect of the meshIO api where it is always handy to write out very basic glb meshes (intended primarily to support debugging cases), but would preserve the current features of assimp support when it is available?  Looking over the meshIO code the comments indicate the assimp support is "primarily intended as a demonstration" which makes me think maybe one of the minimalist gltf readers and writers could provide enough to be useful?

I'd consider giving it a go and submitting a patch, but wanted to check how such a thing might be received before diving in.

**Comment by elalish** [comment]
I think that's a good idea. I'd be especially supportive if you felt like adding our `EXT_mesh_manifold` glTF extension so that we could losslessly roundtrip the data (I already did this in TS) - with that I'd be willing to drop Assimp entirely; the only formats that have any value for manifoldness are glTF and 3MF. Mainly I want to avoid bloating our various binaries, increasing our dependency count, or being on the hook for file format maintenance. To have proper glTF support, we'd need to support textures, but that also means having to keep up with the glTF spec. 

I'm open to ideas, but somehow we have to set expectations, both for our level of maintenance and for what features work. With my TS, all I need to do is update our dependency on `gltf-transform` to pull in all the latest, and no one is sad about an npm dependency. Not so with C++!

**Comment by starseeker** [comment]
Do I understand correctly that the assimp I/O currently in place does adequately support the ```EXT_mesh_manifold``` extension?  If so then I'd imagine it'll be able to figure it out - I'll look into it.

**Reply by elalish** [reply]
No, Assimp has none of that. There's TS code for it here though:  https://github.com/elalish/manifold/blob/master/bindings/wasm/examples/manifold-gltf.ts

**Reply by starseeker** [reply]
OK.  I'll see what I can figure out and submit a PR once I have something worth discussing.

**Reply by starseeker** [reply]
This is the correct/current extension documentation? https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/EXT_mesh_manifold/README.md

**Reply by elalish** [reply]
Yes, that's right.

**Reply by butcherg** [reply]
A bit necro, but it's a related question.  Why not just re-wicker meshIO to do only gltf and lib3mf? I don't understand the reticence regarding assimp, but I get it, it's not my software. I'm using meshIO right now to try getting 3mf to work; it seems it should but I can't create a .3mf file that 3D builder likes.  All meshIO/assimp-produced 3mf get the the "error code: 0x80511002" complaint.

I've put aside the simple CAD project for the time being, currently working on making stone textures for a railroad station project.  Wrote a franken-program that uses opencv contours routines, hmm heightmap->mesh routines, and manifold put-it-all-together code, would be nice to just have 3mf export work...

**Reply by elalish** [reply]
Agreed - it's working great with the JS bindings. If you'd like to contribute on the format side, you're more than welcome.

**Reply by butcherg** [reply]
It's not clear to me - are you going to remove the assimp dependency from manifold?

I'll pick at 3mf a bit, but my priority is to make print files the railroad station.  So, C++ and stl it is, for that. I've already pulled meshIO.h/cpp into my source tree; if I do make 3mf hay I'll contribute the PR, if they're still in your tree.

**Reply by elalish** [reply]
No immediate plans, but I'm open to the idea.

**Reply by starseeker** [reply]
My original question is OBE, since the latest releases have a built-in capability to read and write debugging mesh data via custom OBJ files without any extra dependencies.

**Reply by elalish** [reply]
Yeah, and to be clear, manifold has never depended on assimp, only MeshIO, which is mostly a debug tool.
